rules.d/60-persistent-storage.rules.in

```diff
@@ -161,6 +161,17 @@ ENV{ID_PART_ENTRY_SCHEME}=="gpt", ENV{ID_PART_ENTRY_NAME}=="?*", SYMLINK+="disk/
 # the right device, to access specific disks in a race-free fashion)
 ENV{DISKSEQ}=="?*", ENV{ID_IGNORE_DISKSEQ}!="1", SYMLINK+="disk/by-diskseq/$env{DISKSEQ}$env{.PART_SUFFIX}"
 
+# by-parttype links (Discoverable Partitions Spec)
+ENV{ID_PART_ENTRY_SCHEME}=="gpt", ENV{ID_BOOT_DISK}=="1", ENV{ID_FS_USAGE}=="filesystem|other|crypto", GOTO="persistent_storage_by_parttype_start"
+GOTO="persistent_storage_by_parttype_end"
+LABEL="persistent_storage_by_parttype_start"
+ENV{ID_PART_ENTRY_NAME}=="?*", ENV{.PART_NAME}="$env{ID_PART_ENTRY_NAME}"
+ENV{.PART_NAME}=="", ENV{ID_FS_LABEL_ENC}=="?*", ENV{.PART_NAME}="$env{ID_FS_LABEL_ENC}"
+ENV{.PART_NAME}=="", ENV{ID_PART_ENTRY_UUID}=="?*", ENV{.PART_NAME}="$env{ID_PART_ENTRY_UUID}"
+ENV{.PART_NAME}=="", GOTO="persistent_storage_by_parttype_end"
+ENV{ID_PART_ENTRY_TYPE}=="?*", SYMLINK+="disk/by-parttype/$env{ID_PART_ENTRY_TYPE}/$env{.PART_NAME}"
```

Comment:
this symlink should probably carry "boot" in the name somewhere.

And I think it would make sense to patch block_id, to look up the id we maintain for partition type uuids and simply append it as yet another property onto the device. Then use that property here to create a named symlink.

That said, for compat I think we should always create the uuid-based symlink, and the id-based symlink if we have one. That way, people can use the uuid always without having to rely us to register the part type they need one day.

also, this should initialize link_priority from the partno somehow, so that we get the guarantee that the first partitoin always gets the symlink if multiple partitions have the same type on the same disk.

Reply:
hmm, so i was thinking the symlink wouldn't also contain the label, but just the parttype. i.e. that you have a stable link to the first partition of a specific type, and the label doesn't matter. e.g. a stable reference to ESP, XBOOTLDR, swap and other partitions could be maintained that way, that just always works, and where multiple partitions of the same type on the same disk doesn't make sense.

but your usecase seems different if you want to include the label too. can you elaborate?

Reply:
Basically the idea is that for some partitions, we do want multiple of them (i.e. we'll have two /usr trees). Including the label lets you access all the partitions on the disk

I agree that it doesn't make sense for many partition types (i.e. swap, /var, etc). However, the idea is that this mechanism is also generic, not just for partition types defined by systemd/uapi-group. So if Vanilla OS wants to define `08e5611e-7996-4a0b-b793-0cb63b2a3e14 = grub installation`, and they have two grub installations, they'd be able to find both installations of grub in `/dev/disk/by-parttype/08e5611e-7996-4a0b-b793-0cb63b2a3e14/*` (this is not a contrived example. I'm pretty sure that Vanilla does have grub installed 2x)

However, perhaps picking the first partition we find on the disk might be a good tradeoff. Anyone who needs anything more than that should dissect the GPT table and find the partition themselves?

Something else to consider: sometimes we don't want the first partition we find. Sometimes we want the latest partition by version number

Reply:
hmm, we could simply add both kinds of symlinks i guess. i.e. with the label, and without it. i think that might be quite nice actually. key is that we format them into a strong that has a clear delimiter where the label begins so that there could never be confusion. and we should suppress the additional with-label symlinks for cases where the label is empty anyway.